### ISSUE TITLE
Feature: Row IDs

### ISSUE THREAD

**Issue by westonpace** [lancedb/lancedb#1120]
### SDK

Rust

### Description

In almost all cases there is significant benefit in storing some kind of stable "row id" with rows.  Users can do this manually today (just create a column named row_id) and many users already have a row_id column.  However, when users do not, then maintaining these row ids becomes a burden.

LanceDb should automatically generate a row id column when users do not already have a row id column.  These can either be populated via some kind of UUID or through some kind of "sequence" concept.  Users should be able to mark (in some way) that they already have a row id column and, if so, LanceDb should not generate anything but use the existing row id column.

Note: Lance has an internal concept of a "row address".  However, these are not stable (they change on compaction and may change on other operations in the future) and should not really be exposed.

**Comment by wjones127** [comment]
There is some possibility we might introduce a row id in Lance. One candidate is using the original insertion location as the id: `<64-bit version added><32-bit fragment id><32-bit fragment offset>`. That gives us a 128-bit id (same size as UUID), which guarantees no collisions. It would be implicit (no storage required) when first inserted, and only be explicit on the first time it moved.

It's TBD whether we will need this. But there might be some features in Lance that require it.

**Comment by wjones127** [comment]
> Using this concept, would the row_id also carry uniqueness semantics (i.e 2 records having the same row_id wouldn't be possible) ?

Yes, we will have a unique `row_id` that is an auto incrementing u64 id. It will always be unique between two rows.
